x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java

```diff
@@ -0,0 +1,376 @@
+     * {@link SystemUser}</li>
+     * <li>anonymous access is enabled and this will be considered an anonymous request</li>
+     * </ul>
+     * <p>
+     * Regardless of the scenario, this method will call the listener with either failure or success.
+     */
+    // TODO: In theory, handleNullToken should not be called at all if any credentials is extracted
+    // pkg-private for tests
+    void handleNullToken(Authenticator.Context context, ActionListener<Authentication> listener) {
+        final Authentication authentication;
+        if (context.getFallbackUser() != null) {
+            // TODO: assert we really haven't extract any token
+            logger.trace(
+                "No valid credentials found in request [{}], using fallback [{}]",
+                context.getRequest(),
+                context.getFallbackUser().principal());
+            Authentication.RealmRef authenticatedBy = new Authentication.RealmRef("__fallback", "__fallback", nodeName);
+            authentication = new Authentication(context.getFallbackUser(),
+                authenticatedBy,
+                null,
+                Version.CURRENT,
+                Authentication.AuthenticationType.INTERNAL,
+                Collections.emptyMap());
+        } else if (shouldFallbackToAnonymous(context)) {
+            logger.trace(
+                "No valid credentials found in request [{}], using anonymous [{}]",
+                context.getRequest(),
+                anonymousUser.principal());
+            Authentication.RealmRef authenticatedBy = new Authentication.RealmRef("__anonymous", "__anonymous", nodeName);
+            authentication = new Authentication(anonymousUser,
+                authenticatedBy,
+                null,
+                Version.CURRENT,
+                Authentication.AuthenticationType.ANONYMOUS,
+                Collections.emptyMap());
+        } else {
+            authentication = null;
+        }
+
+        if (authentication != null) {
+            // TODO: we can also support run-as for fallback users if needed
+            // TODO: the authentication for fallback user is now serialised in the inner threadContext
+            //       instead of at the AuthenticationService level
+            writeAuthToContext(context, authentication, listener);
+        } else {
+            final ElasticsearchSecurityException ese = context.getRequest().anonymousAccessDenied();
+            if (false == context.getUnsuccessfulMessages().isEmpty()) {
+                logger.debug("Authenticating with null credentials is unsuccessful in request [{}]" +
+                    " after unsuccessful attempts of other credentials", context.getRequest());
+                final ElasticsearchSecurityException eseWithPreviousCredentials =
+                    new ElasticsearchSecurityException("unable to authenticate anonymously", ese.status(), ese.getCause());
```

Comment:
I changed it to `unable to authenticate as the anonymous user`.

Strictly speaking, it should be  "the anonymous user or the fallback user". But it is more confusing than helpful to metion the fallback user because it is an internal concept and this exception goes back to end-user.